pkg/package-server/provider/registry.go

```diff
@@ -0,0 +1,472 @@
+		defer p.mu.Unlock()
+
+		for i, client := range p.clients {
+			if client.source.GetName() == catsrc.GetName() && client.source.GetNamespace() == catsrc.GetNamespace() {
+				client.conn.Close()
+				p.clients = append(p.clients[:i], p.clients[i+1:]...)
+				logger.Debug("removed gRPC connection")
+				return true
+			}
+		}
+		return false
+	}()
+
+	if !found {
+		logger.Errorf("gRPC connection not found")
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultEventTimeout)
+	defer cancel()
+	p.publishEvents(ctx, logger, catsrc.GetNamespace(), pkgs.Items, deletePkg)
+}
+
+func (p *RegistryProvider) Get(namespace, name string) (*v1alpha1.PackageManifest, error) {
+	logger := logrus.WithFields(logrus.Fields{
+		"action":    "Get PackageManifest",
+		"name":      name,
+		"namespace": namespace,
+	})
+
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	for _, client := range p.clients {
+		if client.source.GetName() == name && client.source.GetNamespace() == namespace || client.source.GetNamespace() == p.globalNamespace {
+			pkg, err := client.GetPackage(context.Background(), &api.GetPackageRequest{Name: name})
+			if err != nil {
+				return nil, err
+			}
+			return toPackageManifest(pkg, client)
+		}
+	}
+	logger.Error("package not found")
+
+	return nil, fmt.Errorf("package %s not found in namespace %s", name, namespace)
+}
+
+func (p *RegistryProvider) List(namespace string) (*v1alpha1.PackageManifestList, error) {
+	logger := logrus.WithFields(logrus.Fields{
+		"action":    "List PackageManifests",
+		"namespace": namespace,
+	})
+
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	pkgs := []v1alpha1.PackageManifest{}
+	for _, client := range p.clients {
+		if client.source.GetNamespace() == namespace || client.source.GetNamespace() == p.globalNamespace || namespace == "" {
+			logger.Debugf("found CatalogSource %s", client.source.GetName())
+
+			stream, err := client.ListPackages(context.Background(), &api.ListPackageRequest{})
+			if err != nil {
+				logger.Warnf("error getting stream: %v", err)
+				return nil, err
+			}
+			for {
+				logger.Info("in for loop getting data from list stream...")
+				pkgName, err := stream.Recv()
+				if err == io.EOF {
+					logger.Info("end of stream")
+					break
+				}
+
+				if err != nil {
+					logger.Warnf("some error while getting data: %v", err)
+					return nil, err
+				}
+				pkg, err := client.GetPackage(context.Background(), &api.GetPackageRequest{Name: pkgName.GetName()})
+				if err != nil {
+					logger.Warnf("some error while getting package: %v", err)
+					return nil, err
+				}
+				newPkg, err := toPackageManifest(pkg, client)
+				pkgs = append(pkgs, *newPkg)
+			}
+		}
+	}
+
+	return &v1alpha1.PackageManifestList{Items: pkgs}, nil
+}
+
+func (p *RegistryProvider) forward(ctx context.Context, logger *logrus.Entry, out chan v1alpha1.PackageManifest, transform forwarder.Transform) {
+	from := make(chan interface{})
+	to := make(chan interface{})
+
+	go func() {
```

Comment:
isn't this goroutine just an extra `forwarder` layer?

Couldn't you just have:

```go 
func (p *RegistryProvider) forward(ctx context.Context, logger *logrus.Entry, out chan v1alpha1.PackageManifest, transform forwarder.Transform) {
  go func() {
    p.forwarder.Forward(ctx, p.pkgTopic, out, transform)
    // ...
  }
}
```
or am I missing some other purpose?

Reply:
Unfortunately we need intermediate channels to convert between `interface {}` and `PackageChan`. The `FanoutForwarder` is still useful in that it ensures the 1:n relationship, but if you want I can forego the abstraction and make a `PackageChan` typed implementation of the pattern - then we wouldn't need this step.

Reply:
Just make a note of it (maybe just a `TODO` comment) and don't worry about it for this PR